### DISCUSSION TITLE
Render filter blocks with Interactivity API

### DISCUSSION THREAD

**Discussion by dinhtungdu** [woocommerce/woocommerce#45038] [WooCommerce, Blocks, and Site Editing]
We’re working on new filter blocks. We choose to integrate Interactivity API into those blocks because it solves the performance issue on the front end nicely. However, we also encountered an implementation issue with the API. In this post, we’re discussing the code duplication issue with blocks utilizing Interactivity API and some attempts we made to solve it.

  

## The current approach and its duplication issue

Filter blocks are dynamic blocks, so for the front end, we render blocks using PHP. But we also want an interactive editor experience, so as with other blocks, we use React for the editor. As a result, we have to maintain two similar components. For complex blocks like attribute filters, the maintenance issue becomes bigger.

  

This is not a new issue though, it’s been around since we had dynamic blocks. But Interactivity API gives us some ideas to solve the duplication issue. We tried and failed.

  

### Saving markup to database

Our first attempt is to use the `<Save>`  component as the single source of truth. In [this example](https://github.com/woocommerce/woocommerce-blocks/blob/f07d684f1c7577b78881221e41e6ea0c9879e8e8/assets/js/blocks/collection-filters/inner-blocks/price-filter/save.tsx "https://github.com/woocommerce/woocommerce-blocks/blob/f07d684f1c7577b78881221e41e6ea0c9879e8e8/assets/js/blocks/collection-filters/inner-blocks/price-filter/save.tsx"), both `<Save>`  and `<Edit>`  components share the same content. For the front end, we use `WP_HTML_Tag_Processor`  to [inject](https://github.com/woocommerce/woocommerce-blocks/blob/f07d684f1c7577b78881221e41e6ea0c9879e8e8/src/BlockTypes/CollectionPriceFilter.php#L67-L83 "https://github.com/woocommerce/woocommerce-blocks/blob/f07d684f1c7577b78881221e41e6ea0c9879e8e8/src/BlockTypes/CollectionPriceFilter.php#L67-L83") the dynamic value to the markup before rendering. In doing so, we only have to maintain one component, we can also reuse the current React components library.

But it has a dealbreaker for us: when we update the blocks, the merchant needs to update the templates/pages containing our filters block to get the latest version. This forces us to abandon this direction.

### Using a template engine

We need to share the markup between JS and PHP renders callback to solve the duplication issue, which sounds like a legitimate use-case of the template engine. We tried `mustache` and it works (see [https://github.com/woocommerce/woocommerce-blocks/pull/10871](https://github.com/woocommerce/woocommerce-blocks/pull/10871)):

- We only need to maintain the mustache template (in the example PR, I embedded it inside the block class, but it’s possible to extract the template into an HTML file).
- Mustache has official implementations for both JS and PHP.

But obviously, it adds another overhead to our already busy stack. @sunyatasattva also points out that with the template engine, we also can’t reuse our existing component library, which in some sense is a duplication issue.

## In closing

We’re using the official approach to keep the project going. But we’re actively looking for a better solution. What do you think about these above attempts? Do you have other ideas to solve this issue?

**Comment by nerrad** [comment]
> We’re using the official approach to keep the project going. But we’re actively looking for a better solution. What do you think about these above attempts? Do you have other ideas to solve this issue?

I really appreciate the thought going into this and looking for ways to reduce duplication of components. A while back, I explored something similar to the [template engine approach you suggested](https://github.com/nerrad/hybrid-block-demo). 

This sounds like a problem that should be solved at the WP level and we should watch for an opportunity to collaborate upstream on this - but for _now_ we continue with the "official approach".
